### PR TITLE
Ensure that all post-build checks have skip policies.

### PR DESCRIPTION
**Pull request by BillyONeal**
<s>Depends on https://github.com/microsoft/vcpkg/pull/38629</s>
<s>Depends on https://github.com/microsoft/vcpkg-docs/pull/319</s>

Also audited for:
* Consistently following the 'IDE error format' pointing to `portfile.cmake`.
* Consistently referring to the correct skip policy in the error message where applicable.
* Printing paths and other information relative to `${CURRENT_PACKAGES_DIR}` or similar as appropriate.
* Debug before release.
* Don't duplicate messages when a problem happens for both debug and release.

Drive by bugfix:

```c++
    if (!build_info.policies.is_enabled(BuildPolicy::ONLY_RELEASE_CRT))
    {
        error_count += check_crt_linkage_of_libs(
            build_info, false, debug_libs, debug_lib_info.value_or_exit(VCPKG_LINE_INFO), msg_sink);
    }
```

was wrong; `ONLY_RELEASE_CRT` should mean 'these should all be release', not 'we aren't even going to look at debug'.

It is fixed to:

```c++
    error_count += check_crt_linkage_of_libs(build_info.crt_linkage,
                                        build_info.policies.is_enabled(BuildPolicy::ONLY_RELEASE_CRT),
                                              debug_libs,
                                              debug_lib_info.value_or_exit(VCPKG_LINE_INFO),
                                              msg_sink);
```

here.

Drive-by bugfix:

`EMPTY_INCLUDE_FOLDER` should not block `CMAKE_HELPER_PORT` from checking that the include folder is, in fact, empty.

Drive-by bugfix:

```c++
        if (!violations.empty())
        {
            msg_sink.println_warning(msgPortBugRestrictedHeaderPaths);
            print_paths(msg_sink, violations);
            msg_sink.println(msgPortBugRestrictedHeaderPaths);     // <-- duplicate!
            return LintStatus::PROBLEM_DETECTED;
        }
```

Drive-by bugfix:
There were duplicate checks for the `lib/cmake` and `debug/lib/cmake` directories which even printed the same message; these have been merged.

Drive-by bugfix:
The existing copyright check emitted nonsense output that included full absolute paths like:

```console
warning: The software license must be available at ${CURRENT_PACKAGES_DIR}/share/cmdline/copyright. vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/D:/b/cmdline//LICENSE")
```

since at least https://github.com/microsoft/vcpkg-tool/pull/1116/ That check has been basically rewritten to use `vcpkg_install_copyright` if there's only one src directory, and otherwise give up and print them all

Drive-by bugfix:
Several of the outdated CRTs didn't have the debug versions listed. Also, P was duplicated and there was no R for VC6.

<s>Drive-by bugfix:
https://github.com/microsoft/vcpkg-tool/pull/1368/files#diff-9a4355abe0b8aae1ff2cd6cc86c0e4ce6bc4e6b7c469eb41332da0325cd20c88R984 broke usage text for header-only libs by adding an extra paren.</s> This was merged in https://github.com/microsoft/vcpkg-tool/pull/1403

### CODE REVIEW

File: src/vcpkg/postbuildlint.cpp [hunk 1]
```diff
@@ -290,136 +349,228 @@ namespace vcpkg
...
+        std::vector<Path> relative_dlls =
+            fs.get_regular_files_recursive_lexically_proximate(package_dir / prefix, IgnoreErrors{});
+        Util::erase_remove_if(relative_dlls, NotExtensionCaseInsensitive{".dll"});
+        add_prefix_to_all(relative_dlls, prefix);
+        return relative_dlls;
+    }
+
+    static LintStatus check_for_dlls_in_lib_dirs(const ReadOnlyFilesystem& fs,
+                                                 const Path& package_dir,
+                                                 View<StringLiteral> lib_dir_prefixes,
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/postbuildlint.cpp:432]: Yeah I'll move these out to namespace scope and drop the arg when they are constants.

File: src/vcpkg/postbuildlint.cpp [hunk 2]
```diff
@@ -580,87 +743,96 @@ namespace vcpkg
...
+                       .append_raw('\n');
         for (const FileAndArch& b : binaries_with_invalid_architecture)
         {
-            msg_sink.println_warning(LocalizedString().append_indent().append(msgBinaryWithInvalidArchitecture,
-                                                                              msg::path = b.file,
-                                                                              msg::expected = expected_architecture,
-                                                                              msg::actual = b.actual_arch));
+            msg.append_raw(NotePrefix)
+                .append(msgBinaryWithInvalidArchitecture,
+                        msg::path = b.relative_file.generic_u8string(),
```
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/postbuildlint.cpp:772]: I wanted all the paths in these diagnostics to use CMake syntax. (It also has the side benefit of  making writing the tests moderately easier)

[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/postbuildlint.cpp:772]:
```powershell
    $expected = @"
$($PortfilePath): warning: The triplet requests that binaries are built for x86, but the following binaries were built for a different architecture. This usually means toolchain information is incorrectly conveyed to the binaries' build system. To suppress this message, add set(VCPKG_POLICY_SKIP_ARCHITECTURE_CHECK enabled)
$($packagesRoot)$($NativeSlash)test-dll_$($Triplet): note: the binaries are relative to `${CURRENT_PACKAGES_DIR} here
note: debug/lib/test_dll.lib is built for x64
note: lib/test_dll.lib is built for x64
note: debug/bin/test_dll.dll is built for x64
note: bin/test_dll.dll is built for x64
"@
```

File: azure-pipelines/e2e-assets/test-dll-port-template/portfile.cmake [hunk 3]
```diff
@@ -1,14 +1,69 @@
...
-file(COPY "${CMAKE_CURRENT_LIST_DIR}/debug/test_dll.dll" DESTINATION "${CURRENT_PACKAGES_DIR}/debug/bin")
-file(COPY "${CMAKE_CURRENT_LIST_DIR}/debug/test_dll.lib" DESTINATION "${CURRENT_PACKAGES_DIR}/debug/lib")
-file(COPY "${CMAKE_CURRENT_LIST_DIR}/release/test_dll.dll" DESTINATION "${CURRENT_PACKAGES_DIR}/bin")
-file(COPY "${CMAKE_CURRENT_LIST_DIR}/release/test_dll.lib" DESTINATION "${CURRENT_PACKAGES_DIR}/lib")
 file(COPY "${CMAKE_CURRENT_LIST_DIR}/test.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+
+file(MAKE_DIRECTORY "${CURRENT_PACKAGES_DIR}/share/${PORT}")
 file(TOUCH "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright")
+
+if(release-only IN_LIST FEATURES AND NOT bad-release-only IN_LIST FEATURES)
```
Comment (ras0219-msft) [inline azure-pipelines/e2e-assets/test-dll-port-template/portfile.cmake:7]:
```suggestion
if("release-only" IN_LIST FEATURES AND NOT "bad-release-only" IN_LIST FEATURES)
```

I'm not 100% sure that this is a problem (due to the hyphen), but it does make me very nervous about accidental variable expansion.

File: azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1 [hunk 4]
```diff
@@ -1,82 +1,987 @@
...
+if ($buildOutput.Replace("`r`n", "`n").Contains($expected)) {
+    throw 'VCPKG_POLICY_ALLOW_RESTRICTED_HEADERS didn''t allow'
+}
+
+Refresh-TestRoot
+$buildOutput = Run-VcpkgAndCaptureOutput @commonArgs install --overlay-ports="$PSScriptRoot/../e2e-ports" 'vcpkg-policy-include-folder[do-install,do-install-debug]' --no-binarycaching --enforce-port-checks
+Throw-IfNotFailed
+$expected = @"
+$($PortfilePath): warning: `${CURRENT_PACKAGES_DIR}/debug/include should not exist. To suppress this message, add set(VCPKG_POLICY_ALLOW_DEBUG_INCLUDE enabled)
+note: If this directory was created by a CMake build system that does not allow installing headers in debug to be disabled, delete the duplicate directory with file(REMOVE_RECURSE "`${CURRENT_PACKAGES_DIR}/debug/include")
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1:93]:
```suggestion
note: If this directory was created by a build system that does not allow installing headers in debug to be disabled, delete the duplicate directory with file(REMOVE_RECURSE "`${CURRENT_PACKAGES_DIR}/debug/include")
```

File: azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1 [hunk 5]
```diff
@@ -1,82 +1,987 @@
...
-$buildOutput = Run-VcpkgAndCaptureOutput --triplet x64-uwp "--x-buildtrees-root=$buildtreesRoot" "--x-install-root=$installRoot" "--x-packages-root=$packagesRoot" install --overlay-ports="$TestingRoot/wrong-appcontainer" test-dll --no-binarycaching
-$expected = "warning: The App Container bit must be set for Windows Store apps. The following DLLs do not have the App Container bit set:`n" + `
-"`n" + `
-"  $packagesRoot\test-dll_x64-uwp\debug\bin\test_dll.dll`n" + `
-"  $packagesRoot\test-dll_x64-uwp\bin\test_dll.dll`n"
+Refresh-TestRoot
+$buildOutput = Run-VcpkgAndCaptureOutput @commonArgs install --overlay-ports="$PSScriptRoot/../e2e-ports" 'vcpkg-policy-misplaced-cmake-files[do-install-cmake]' --no-binarycaching --enforce-port-checks
+Throw-IfNotFailed
+$expected = @"
+$($PortfilePath): warning: This port installs the following CMake files in places CMake files are not expected. CMake should be installed in `${CURRENT_PACKAGES_DIR}/share/`${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1:131]:
```suggestion
$($PortfilePath): warning: This port installs the following CMake files in places CMake files are not expected. CMake files should be installed in `${CURRENT_PACKAGES_DIR}/share/`${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
```

File: azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1 [hunk 6]
```diff
@@ -1,82 +1,987 @@
...
+}
+if (-not $buildOutput.Replace("`r`n", "`n").Contains($expected)) {
+    throw 'Did not detect bad CMake files'
+}
+
+Refresh-TestRoot
+$buildOutput = Run-VcpkgAndCaptureOutput @commonArgs install --overlay-ports="$PSScriptRoot/../e2e-ports" 'vcpkg-policy-misplaced-cmake-files[do-install-lib]' --no-binarycaching --enforce-port-checks
+Throw-IfNotFailed
+$expected = @"
+$($PortfilePath): warning: This port installs the following CMake files in places CMake files are not expected. CMake should be installed in `${CURRENT_PACKAGES_DIR}/share/`${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1:147]:
```suggestion
$($PortfilePath): warning: This port installs the following CMake files in places CMake files are not expected. CMake files should be installed in `${CURRENT_PACKAGES_DIR}/share/`${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
```

File: azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1 [hunk 7]
```diff
@@ -1,82 +1,987 @@
...
+}
+if (-not $buildOutput.Replace("`r`n", "`n").Contains($expected)) {
+    throw 'Did not detect bad CMake files'
+}
+
+Refresh-TestRoot
+$buildOutput = Run-VcpkgAndCaptureOutput @commonArgs install --overlay-ports="$PSScriptRoot/../e2e-ports" 'vcpkg-policy-misplaced-cmake-files[do-install-cmake,do-install-lib]' --no-binarycaching --enforce-port-checks
+Throw-IfNotFailed
+$expected = @"
+$($PortfilePath): warning: This port installs the following CMake files in places CMake files are not expected. CMake should be installed in `${CURRENT_PACKAGES_DIR}/share/`${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1:164]:
```suggestion
$($PortfilePath): warning: This port installs the following CMake files in places CMake files are not expected. CMake files should be installed in `${CURRENT_PACKAGES_DIR}/share/`${PORT}. To suppress this message, add set(VCPKG_POLICY_SKIP_MISPLACED_CMAKE_FILES_CHECK enabled)
```

File: azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1 [hunk 8]
```diff
@@ -1,82 +1,987 @@
...
+    throw 'VCPKG_POLICY_SKIP_MISPLACED_REGULAR_FILES_CHECK didn''t suppress'
 }
+
+# Misplaced pkgconfig
+Refresh-TestRoot
+$PortfilePath = "$PSScriptRoot/../e2e-ports$($NativeSlash)vcpkg-policy-misplaced-pkgconfig$($NativeSlash)portfile.cmake"
+$buildOutput = Run-VcpkgAndCaptureOutput @commonArgs install --overlay-ports="$PSScriptRoot/../e2e-ports" 'vcpkg-policy-misplaced-pkgconfig[install-arch-dependent-bad-misplaced]' --no-binarycaching --enforce-port-checks
+Throw-IfNotFailed
+$expected = @"
+$($PortfilePath): warning: The following misplaced pkgconfig directories were installed. Misplaced pkgconfig files will not be found correctly by pkgconf or pkg-config. pkgconfig directories should be `${CURRENT_PACKAGES_DIR}/share/pkgconfig (for architecture agnostic / header only libraries only), `${CURRENT_PACKAGES_DIR}/lib/pkgconfig (for release dependencies), or `${CURRENT_PACKAGES_DIR}/lib/debug/pkgconfig (for debug dependencies). To suppress this message, add set(VCPKG_POLICY_SKIP_PKGCONFIG_CHECK enabled)
```
Comment (ras0219-msft) [inline azure-pipelines/end-to-end-tests-dir/post-build-checks.ps1:826]:
```suggestion
$($PortfilePath): warning: The following misplaced pkgconfig directories were installed. Misplaced pkgconfig files will not be found correctly by pkgconf or pkg-config. pkgconfig directories should be `${CURRENT_PACKAGES_DIR}/share/pkgconfig (for architecture agnostic / header only libraries only), `${CURRENT_PACKAGES_DIR}/lib/pkgconfig (for release dependencies), or `${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig (for debug dependencies). To suppress this message, add set(VCPKG_POLICY_SKIP_PKGCONFIG_CHECK enabled)
```

**Review by ras0219-msft — APPROVED** [review]

File: azure-pipelines/e2e-assets/test-dll-port-template/portfile.cmake (hunk 3 above)
Comment (BillyONeal) [reply azure-pipelines/e2e-assets/test-dll-port-template/portfile.cmake:7]: I'm 100% sure this isn't a problem due to IN_LIST but I'm making the change anyway.

**Pull request merged**